solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -1701,7 +1701,7 @@ public void publish(final CoreDescriptor cd, final Replica.State state, boolean
         cd.getCloudDescriptor().setLastPublished(state);
       }
       DocCollection coll = zkStateReader.getCollection(collection);
-      if (forcePublish || sendToOverseer(coll, coreNodeName)) {
+      if (sendToOverseer(coll, coreNodeName)) {
```

Comment:
Can you comment on this change (added with 35a511d46875012da80802c0f4289db4b6d56c20)?

Reply:
Just adding my observation here. `forcePublish` is an old flag that force a publish to overseer even if the core is [closed](https://github.com/fullstorydev/lucene-solr/blame/663684fa05fb538ac3afa3a263cf0644209d9192/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1624).

I don't fully understand why we have added that along with the PRS changes, my guess is that we want the preserve the behavior of "publish to os no matter what" (while the exact behavior when it's both `true` and PRS enabled is not well defined)

The problem with such flag tho, is when  during `preRegister` from ZkController on new cores (which is invoked once per core), it calls `publish(cd, Replica.State.DOWN, false, true);` which `forcePublish` is true here, and since `true` means always send to Overseer, so even for PRS collection, it will publish such `state.json` to overseer n times, which n is the number of core. So from OS point of view, it will receive n messages, which n is number of shards.

This is largely okay for non PRS collection due to the "batching" behavior of `ZkStateWriter` which it will only write the "latest" state.json update (bound by certain refresh interval), hence the actual write ops to ZK is minimal. However, this is NOT okay for PRS collections, as batching in `ZkStateWriter` is essentially disabled for PRS collections, hence n number of writes will be sent to ZK.

Now the removal of such `forcePublish` will work around this problem as it will no longer submit messages for each core registration for PRS collection (hence less Zk writes), however, there are just some items to consider:
1. Probably want to double confirm the logic in `sendToOverseer` as some messages that used to rely on `forcePublish` flag to get pushed to overseer, will now have to run the logic in `sendToOverseer` to determine where the message should go
2. We probably want to better define the behavior of PRS collection here. This is publishing the Core info. If such core does not exist in zk. we might want to do both? (Send to ZK to add entry in state.json AND set PRS entries) . As for core that does exist in zk state.json already, it's probably okay to just set PRS entries. The existing logic seems to want to handle the case that is PRS and replica/core is not yet in ZK (by looking at logic in  `sendToOverseer`), however, it's probably not working well as it does not update the PRS entries in that case. It might not be a big deal now as the state.json for such replica is being inserted in place such as [`CreateCollectionCmd`](https://github.com/fullstorydev/lucene-solr/blob/release/8.8/solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java#L282) Though we still probably want to better define the behavior to cover all cases.

And in longer run, if we still want state.json to keep info of the "existence" of a core/replica, then we really should fix the batching for PRS collection in `ZkStateWriter` and probably eliminate individual direct writes on state.json for PRS collections

Reply:
@noblepaul @chatman I think keep this PR for just `collection creation` issue only. We can revisit this separately.

Reply:
@patsonluk 

>I don't fully understand why we have added that along with the PRS changes, my guess is that we wanted to preserve the behavior of "publish to os no matter what" (while the exact behavior when it's both true and PRS enabled is not well defined)

The original motivation during PRS impl was not to eliminate overseer messages 100% . Even if we eliminated 90% (state messages) it was good enough. In other words, we were trying to be a little conservative (by preserving the old behavior to a little extent) to ensure correctness.

Now we have reached a point where it can to be eliminated 100% for efficiency and safety (from race conditions).


>This is largely okay for non PRS collection due to the "batching" behavior of ZkStateWriter which it will only write the "latest" state.json update (bound by certain refresh interval), hence the actual write ops to ZK is minimal. However, this is NOT okay for PRS collections, as batching in ZkStateWriter is essentially disabled for PRS collections, hence n number of writes will be sent to ZK.
 
This is true. We got rid of batching in case of PRS collections, since messages that come to overseer for PRS collections are fewer and far between

> Probably want to double confirm the logic in sendToOverseer as some messages that used to rely on forcePublish flag to get pushed to overseer, will now have to run the logic in sendToOverseer to determine where the message should go

We should audit sendToOverseer() again and ensure that no messages go to overseer for PRS collections


>We probably want to better define the behavior of PRS collection here. This is publishing the Core info. If such core does not exist in zk. we might want to do both?

This actually was a legacy behavior where a core can just startup and could register itself as a replica by sending a message to overseer. Today, if core info does not exist in state.json, the replicas should not even start.

>And in longer run, if we still want state.json to keep info of the "existence" of a core/replica, then we really should fix the batching for PRS collection in ZkStateWriter and probably eliminate individual direct writes on state.json for PRS collections.
 
Yes,

No messages should go to overseer for PRS collections for state/leader
All modification operations on PRS collections (e.g. CREATE, ADDREPLICA, DELETEREPLICA) should be done as single atomic operations and avoid multiple updates.


>Which the latter could still be a performance issue ....

The perf issue arose from a race condition which caused the CREATE command wait indefinitely for all replicas to be ACTIVE. So, it was actually not a perf bug, but it was a “deadlock”